domains/runtime/evm/src/lib.rs

```diff
@@ -908,49 +958,98 @@ impl_runtime_apis! {
             }
         }
 
-        fn check_transaction_validity(
+        fn check_transaction_validity(uxt: &<Block as BlockT>::Extrinsic,
+            block_number: BlockNumber,
+            block_hash: <Block as BlockT>::Hash) -> Result<(), domain_runtime_primitives::CheckTxValidityError>  {
+            let lookup = frame_system::ChainContext::<Runtime>::default();
+            check_transaction_validity_inner(&lookup, uxt, block_number, block_hash).map(|_| ())
+        }
+
+        fn check_transaction_validity_and_do_pre_dispatch(
             uxt: &<Block as BlockT>::Extrinsic,
             block_number: BlockNumber,
             block_hash: <Block as BlockT>::Hash
         ) -> Result<(), domain_runtime_primitives::CheckTxValidityError> {
             let lookup = frame_system::ChainContext::<Runtime>::default();
-            let maybe_signer_info = extract_signer_inner(uxt, &lookup);
-
-            if let Some(signer_info) = maybe_signer_info {
-                let signer = signer_info.map_err(|tx_validity_error| {
-                    domain_runtime_primitives::CheckTxValidityError::UnableToExtractSigner {
-                        error: tx_validity_error
-                    }
-                })?;
-
-                let tx_validity =
-                    Executive::validate_transaction(TransactionSource::External, uxt.clone(), block_hash);
 
-                tx_validity.map(|_| ()).map_err(|tx_validity_error| {
-                    domain_runtime_primitives::CheckTxValidityError::InvalidTransaction {
-                        error: tx_validity_error,
-                        storage_keys: storage_keys_for_verifying_tx_validity_inner(signer, block_number, Self::extrinsic_era(uxt)),
+            let maybe_signer = check_transaction_validity_inner(&lookup, uxt, block_number, block_hash)?;
+
+            let storage_keys = storage_keys_for_verifying_tx_validity_inner(maybe_signer, block_number, Self::extrinsic_era(uxt));
+
+            let xt = uxt.clone().check(&lookup).map_err(|tx_validity_error| {
+                domain_runtime_primitives::CheckTxValidityError::InvalidTransaction {
+                    error: tx_validity_error,
+                    storage_keys: storage_keys.clone(),
+                }
+            })?;
+
+            let dispatch_info = xt.get_dispatch_info();
+
+            let encoded = uxt.encode();
+            let encoded_len = encoded.len();
+
+            // We invoke `pre_dispatch` in addition to `validate_transaction`(even though the validation is almost same) as that will add the side effect of SignedExtension in the storage buffer
+            // which would help to maintain context across multiple transaction validity check against same
+            // runtime instance.
+            match xt.signed {
```

Comment:
While it is desired to perform a more exhaustive validation here to cover different types of tx and checks, the `storage_keys` only contains a few storages used in the `SignedExtra` check which may not be exhaustive, i.e. if the `pre_dispatch` of a tx requires access to another storage which is not included in the `storage_keys` then we are unable to construct a valid fraud proof. I think this is a broader problem we need to think on cc @vedhavyas

Reply:
@NingLin-P I have found a way. Idea is to use ProofRecorder to get the keys accessed during runtime execution. These keys are not public at the moment. I have opened a PR here: https://github.com/paritytech/polkadot-sdk/pull/2561

Once that is in, to create storage proof, we simply have to execute desired runtime call and get the accessed keys from `ProofRecorder`.